src/interface/viewport.c

```diff
@@ -1493,6 +1493,440 @@ void viewport_park_entrance_paint_setup(uint8 direction, int height, rct_map_ele
+		regs.ax = height/16;
+		regs.edi = RCT2_GLOBAL(0x9E3278, uint32);
+		surfaceElement = RCT2_GLOBAL(0x9E324C, rct_map_element *);
+		
+		//SAVE
+		saved_ecx = regs.ecx;
+		
+		regs.ah = surfaceElement->base_height / 2;
+		regs.cl = regs.al + RCT2_ADDRESS(0x97B4E4, uint8)[regs.edi];
+		regs.ch = regs.ah + RCT2_ADDRESS(0x97B504, uint8)[regs.ebx];
+		regs.al += RCT2_ADDRESS(0x97B4C4, uint8)[regs.edi];
+		regs.ah += RCT2_ADDRESS(0x97B4A4, uint8)[regs.ebx];
+		RCT2_GLOBAL(0x9E3282, uint16) = regs.ax;
+		RCT2_GLOBAL(0x9E328A, uint16) = regs.cx;
+		
+		//RESTORE
+		regs.ecx = saved_ecx;
+	}
+//loc_660953:
+	//callcode_push1(0x660953, saved_esi, &regs); return;
+	regs.esi = RCT2_GLOBAL(RCT2_ADDRESS_CURRENT_ROTATION, uint32);
+	x = RCT2_GLOBAL(0x9DE568, uint16) + RCT2_ADDRESS(0x97B494, uint16)[regs.esi*2];
+	y = RCT2_GLOBAL(0x9DE56C, uint16) + RCT2_ADDRESS(0x97B496, uint16)[regs.esi*2];
+	RCT2_GLOBAL(0x9E3240, rct_map_element *) = NULL;
+	if(x < 0x2000 && y < 0x2000)
+	{
+		rct_map_element *surfaceElement = map_get_surface_element_at(x/32, y/32); //surfaceElement = esi
+		
+		RCT2_GLOBAL(0x9E3240, rct_map_element *) = surfaceElement;
+		
+		RCT2_GLOBAL(0x9E3254, uint32) = ((surfaceElement->type & 3) << 3) | (surfaceElement->properties.surface.terrain >> 5);
+		
+		regs.ebx = helper(surfaceElement, regs.cl);
+		RCT2_GLOBAL(0x9E3268, uint32) = regs.ebx;
+		regs.ax = height/16;
+		regs.edi = RCT2_GLOBAL(0x9E3278, uint32);
+		surfaceElement = RCT2_GLOBAL(0x9E3240, rct_map_element *);
+		
+		//SAVE
+		saved_ecx = regs.ecx;
+		
+		regs.ah = surfaceElement->base_height / 2;
+		regs.cl = regs.al + RCT2_ADDRESS(0x97B4A4, uint8)[regs.edi];
+		regs.ch = regs.ah + RCT2_ADDRESS(0x97B504, uint8)[regs.ebx];
+		regs.al += RCT2_ADDRESS(0x97B4C4, uint8)[regs.edi];
+		regs.ah += RCT2_ADDRESS(0x97B4E4, uint8)[regs.ebx];
+		RCT2_GLOBAL(0x9E327C, uint16) = regs.ax;
+		RCT2_GLOBAL(0x9E3284, uint16) = regs.cx;
+		
+		//RESTORE
+		regs.ecx = saved_ecx;
+	}
+//loc_660A3C:
+	//RESTORE
+	regs.esi = saved_esi;
+	
+	//ToDo: refactor this
+	//RCT2_CALLFUNC_Y(0x660A3D, &regs); return;
+	if(!(RCT2_GLOBAL(RCT2_ADDRESS_CURRENT_VIEWPORT_FLAGS, uint16) & VIEWPORT_FLAG_LAND_HEIGHTS))
+		goto loc_660AAB;
+	if(RCT2_GLOBAL(0x9E3296, uint16)!=0)
+		goto loc_660AAB;
+	//RCT2_CALLFUNC_Y(0x660A52, &regs); return;
+	
+	//SAVE
+	saved_ecx = regs.ecx;
+	saved_edx = regs.edx;
+	saved_esi = regs.esi;
+	
+	regs.ax = RCT2_GLOBAL(0x9DE56A, uint16) + 0x10;
+	regs.cx = RCT2_GLOBAL(0x9DE56E, uint16) + 0x10;
+	regs.dx = map_element_height(regs.ax, regs.cx) + 3;
+	regs.ebx = regs.dx;
+	regs.ebx = (unsigned)regs.ebx >> 4;
+	regs.ebx += 0x20781689;
+	regs.bx += RCT2_GLOBAL(0x9AACBD, uint16);
+	regs.bx -= RCT2_GLOBAL(0x1359208, uint16);
+	
+	regs.al = 0x10;
+	regs.cl = 0x10;
+	regs.di = 1;
+	regs.si = 1;
+	regs.ah = 0;
+	regs.ebp = RCT2_GLOBAL(RCT2_ADDRESS_CURRENT_ROTATION, uint32);
+	
+	//I think the function being called returns void, but I can never be absolutely sure.
+	RCT2_CALLPROC_X(
```

Comment:
This is one of the main viewport drawing functions it has the exact same parameters  as `sub_98197C` it can return a flag to indicate if its successful but it doesn't do anything to the registers. `ebx` is the `image_id`.